sharktank/sharktank/layers/rotary_embedding.py

```diff
@@ -31,6 +33,48 @@ def __init__(
     def forward(self, *, xq: torch.Tensor, xk: torch.Tensor, start_index: int):
         # xq_, xk_ shape: bs, sl, _, dim
         # freqs_cis shape: max_sl, dim
+
+        def create_interleaved_tensor(dim):
+            """Creates a tensor which indexes an tensor such that
+            it alternates between elements of its first and second
+            half. Intended for use for HuggingFace's rotation
+            implementation.
+
+            Args:
+              dim: Size of tensor
+
+            Returns:
+              Interleaved indexing tensor
+            """
+            first_half = torch.arange(dim // 2)
+            second_half = torch.arange(dim // 2, dim)
+
+            interleaved_tensor = torch.empty(dim, dtype=torch.long)
+            interleaved_tensor[0::2] = first_half
+            interleaved_tensor[1::2] = second_half
+
+            return interleaved_tensor
+
+        def create_ordering_tensor(dim):
+            """Creates a tensor which indexes an tensor such that
+            it reverses the alternation induced by create_interleaved_tesnor.
+            Intended for use for HuggingFace's rotation implementation.
+
+            Args:
+              dim: Size of tensor
+
+            Returns:
+              Ordering indexing tensor
+            """
+            order_tensor = torch.empty(dim, dtype=torch.long)
+            order_tensor[: dim // 2] = torch.arange(0, dim, 2)
+            order_tensor[dim // 2 :] = torch.arange(1, dim, 2)
+            return order_tensor
+
+        if self.use_hf:
+            xq = xq[..., create_interleaved_tensor(xq.shape[-1])]
+            xk = xk[..., create_interleaved_tensor(xq.shape[-1])]
```

Comment:
Not too much context here, but can we (trivially) move this xq/xk into the other `if self.use_hf:`  region?